### ISSUE TITLE
Serialization of machines

### ISSUE THREAD

**Issue by olivierlabayle** [JuliaAI/MLJSerialization.jl#15]
Hi,

I am currently looking at the serialization of machines and particularly machines of composite models. It seems to me the main caveat is that machines store data that we may not want to store on disk. 
For now, I have the following simple recursive approach that unfortunately modifies the instance in place:

```julia
function forget!(mach)
    if mach isa Machine{<:Composite}
        for submach in fitted_params(mach).machines
            forget!(submach)
        end
    end
    # Erase the data in base field and cache
    mach.data = ()
    # It is not possible to change the cache.data field directly as there is a type mismatch
    if mach.cache !== nothing
        mach.cache = Base.structdiff(mach.cache, NamedTuple{(:data,)})
    end
    return mach
end

# Which can be used as follows
forget!(mach)
# And then use any serialization facility, I see you rely on the JLSO library
serialize("mymachine.jls", mach)
saved_mach = deserialize("mymachine.jls")
```

The advantage is that both `fitted_params` and `report` will continue to work afterwards.

I am wondering about your future plans for Serialization, because I think this is highly interconnected with how machines work under the hood (difference between the various fields like data, cache, report, ...). I am happy to discuss your view and happy to contribute if I can.

**Comment by ablaom** [comment]
@olivierlabayle Thanks for looking into this and the offer of help!

Looks like you have given this some thought. I, for one, would be very happy for you to run with something that addresses the following issues around serialization of which I am aware:

1. #16  - an issue I just transferred from MLJ, which you may not have seen
2. If a machine in a learning network is caching model-specific representations of data (`isa Machine{<:Any, true}`) - which is the case for the default constructor - then there is training data attached to that machine that current data anonymization logic does not address. 

Perhaps you have others to add to this?

Recursively removing data sounds like a good idea. But, yeah, mutating the machine is not ideal. If possible, it would be nice if data anonymization happened automatically (without the user explicitly needing to do it). On the other hand, the current logic for handling this (imperfectly) is not that pretty.

**Comment by olivierlabayle** [comment]
Thanks for the pointers!

From what I understand, the required features are as follows:

- Supporting non julia dependencies in the `fitresults` which is currently done by the `save` method to be overloaded.
- Supporting Composite models
- Discarding data

I have started to work on something today but before I dive too deep in anything I'd like to give a brief overview of my intended procedure because it's likely I've missed something:

1. From a given machine build a "copy" that has the following differences with the original machine

   - No remaining data in any of the machine/submachines, concerned fields: `data`, `cache.data`, `resampled_data` and maybe `old_rows`?
   - A serializable fitresult for any of the machine/submachines, concerned field: `fitresult`

The process will deal with `Composite` machines by investigating the `report` field as I think it is where submachines are stored and proceed recursively. 

To deal with TunedModel, EnsembleModel, ... Unless I am missing something, I think it is sufficient to implement the `save` method in the respective repositories.

2. Then, hopefully, any serialization method should work. eg: `serialize(filename, copymach)`

3. The `restore` will only convert recursively the `serializable_fitresult` back to the original as already the case.

How does that look to you? 

In case you have time to look into it a bit, the "pseudocode" for the idea is [below this line](https://github.com/olivierlabayle/MLJSerialization.jl/blob/fb49a409a14339665580e74eb68e11158b92c161/src/machines.jl#L118) and the entry point is `saveable_machine`

**Comment by ablaom** [comment]
Yes, I think you have a pretty good grasp of the problem and the solution looks sound. My feeling is that this is quite a substantial project and we may want to break it down a bit?

1. Data anonymization
2. Serialisation playing well with composition in presence of non-julia components

Step 1 is probably the easier one. 

 > The process will deal with Composite machines by investigating the report field as I think it is where submachines are stored and proceed recursively.

Both `report(composite_mach)` and `fitted_params(composite_mach)` have access points. 

> In case you have time to look into it a bit, the "pseudocode" for the idea is below this line and the entry point is saveable_machine

Will try to look over this today (leaving for a week's leave tomorrow).

**Comment by ablaom** [comment]
Okay, I've had a closer look at the pseudocode and proposal and I think I get this better now. 

So the idea, which I like, is that we decouple the actual serialization step from the de-anonymization  and "persistentization" steps? We add a new method, exposed to the user, called `decontaminate(mach)` or whatever, which applies just the de-anonymization  and "persistentization"  steps, but without mutating `mach`. The user takes over responsibility for serialization using her preferred method. But this seems to imply the user also needs to take over responsibility for restoring the original "non-persistent" representation of the `fitresult`? Or am I missing something? I didn't quite understand your point 3. 

I didn't really understand [this line](https://github.com/olivierlabayle/MLJSerialization.jl/blob/fb49a409a14339665580e74eb68e11158b92c161/src/machines.jl#L128), which relates the to model wrappers. Maybe you can say a bit more about that.

**Comment by olivierlabayle** [comment]
Yes that's my idea but not as easy as I initially thought.

- I have just realised that machines originally live within nodes in the graph, so that's where the change should actually take place. And then only regenerating the report field as done in `return!`. This makes the operation more difficult to do not in place. Maybe [this](https://github.com/JuliaAI/MLJBase.jl/blob/207ce3790485cedc02dfc7540b542759fd94da28/src/composition/learning_networks/machines.jl#L431) could prove useful.

- For the restoration function I was just pointing out that there may not be much new logic beyond what is currently done.

- For wrappers, I haven't played a lot with TunedModel and EnsembleModel etc... But I thought I understood the problem with non julia code was that the `save` function of the underlying model was not called. I was just thinking we should overload `save` for each of those special composite that will in turn call the `save` function of the base model as a subpart of the procedure?

**Comment by ablaom** [comment]
Closed by changes released at MLJBase 0.20.
